### PR TITLE
[Multiple Datasources] Add TLS configuration for multiple data sources

### PR DESCRIPTION
**Pull request by cwperks**
### Description

This PR adds 2 settings to multiple datasources:

```
data_source.ssl.verificationMode
data_source.ssl.certificateAuthorities
```

Example multi-datasource config:

```
data_source.enabled: true
data_source.ssl.verificationMode: full
data_source.ssl.certificateAuthorities: [ "/Users/cwperx/Projects/opensearch/OpenSearch-Dashboards/config/root-ca.pem" ]
```

These settings are common across all configured datasources so setting `verificationMode` to `none` will disable ssl verification for all datasources. Similarly, the list of all certificateAuthorities will be used to verify across all configured datasources. 

These configuration values are separate then their equivalents when OSD is configured with one OpenSearch cluster where these are used to configure the legacy client to connect with the default OpenSearch cluster:

```
opensearch.ssl.verificationMode
opensearch.ssl.certificateAuthorities
```

### Issues Resolved

- https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4524

## Testing the changes

I have been testing by running 2 clusters locally, both configured with the demo certificates. One cluster is the default OpenSearch and the other cluster I add as another data source.

Tested following scenarios:

1. Both clusters have the security plugin installed and https enabled
     - Able to connect to Cluster 2 when `verificationMode` is set to `none`
     - Able to connect to Cluster 2 when `verificationMode` is set to `certificate` and `certificateAuthorities` are set with correct CAs
     - Able to connect to Cluster 2 when `verificationMode` is set to `full` and `certificateAuthorities` are set with correct CAs
     - Not able to connect to Cluster 2 when `verificationMode` is set to `certificate` and `certificateAuthorities` are set with incorrect CAs
     - Not able to connect to Cluster 2 when `verificationMode` is set to `full` and `certificateAuthorities` are set with incorrect CAs

2. Disable HTTPS on cluster 2
    - Able to connect to cluster 2 with https disabled. When entering hostname make sure the protocol is http instead of https

I am setting this PR to draft while tests are added in the security plugin to verify all of these scenarios.

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by seraphjiang** [comment]
We might add the new config into default opensearch_dashboards.yml and comment out with recommended default value

**Comment by cwperks** [comment]
> We might add the new config into default opensearch_dashboards.yml and comment out with recommended default value

@seraphjiang Added commented out examples with reasonable defaults to opensearch_dashboards.yml

**Comment by cwperks** [comment]
@seraphjiang In what circumstances is the [legacy_client](https://github.com/opensearch-project/OpenSearch-Dashboards/tree/main/src/plugins/data_source/server/legacy) used? I'm looking through the data-source plugin now and it looks like this PR may also need to be extended to the legacy client. Is the legacy client used to connect to ES clusters?

**Comment by cwperks** [comment]
> Is the legacy client used to connect to ES clusters?

I see an open issue around [version decoupling](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6177). I added the legacy client changes into this PR as well.

@BionIT @ZilongX Could you take a look at this PR when you have a chance?

File: config/opensearch_dashboards.yml
Comment (ZilongX) [inline config/opensearch_dashboards.yml:289]: do we support only two verificationMode [full, none] or more ?

Comment (ZilongX) [reply config/opensearch_dashboards.yml:289]: kk find three support modes [none, full, certificate], maybe let's add more description in the yml entry comment

File: src/plugins/data_source/server/client/client_config.test.ts
Comment (ZilongX) [inline src/plugins/data_source/server/client/client_config.test.ts:33]: looks like we need another test case `test ssl config with verification mode set to certificate`, also curious what's the difference between mode `full` and `certificate`

File: src/plugins/data_source/server/client/client_config.ts
Comment (ZilongX) [inline src/plugins/data_source/server/client/client_config.ts:46]: ok, so the certificate mode would skip the cert validation whether the certificate cert is issued to hostname - https://nodejs.org/api/tls.html#tlscheckserveridentityhostname-cert

**Comment by ZilongX** [comment]
also triggered rerunning of failed checks, there were some reverted changes in the FT repo yesterday and now all CI should pass

File: src/plugins/data_source/server/client/client_config.test.ts
Comment (cwperks) [reply src/plugins/data_source/server/client/client_config.test.ts:33]: I will add a test for this case

File: src/plugins/data_source/server/client/client_config.ts
Comment (cwperks) [reply src/plugins/data_source/server/client/client_config.ts:72]: This code is based on a similar method in [src/core/server/opensearch/opensearch_config.ts](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/opensearch/opensearch_config.ts#L364-L425) but only supports `certificateAuthorities` as a setting. 

The lines in this function are hit through the tests for `verificationMode = full`, but I can also add tests specifically for this method too. The code in `opensearch_config.ts` is [tested similarly](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/opensearch/opensearch_config.test.ts) as this PR is.

File: src/plugins/data_source/server/client/client_config.test.ts
Comment (cwperks) [reply src/plugins/data_source/server/client/client_config.test.ts:33]: Added test for `certificate` mode that asserts that it overwrites `checkServerIdentity`

File: src/plugins/data_source/server/client/client_config.ts
Comment (cwperks) [reply src/plugins/data_source/server/client/client_config.ts:69]: Any suggestions on where this can be extracted to?

File: config/opensearch_dashboards.yml
Comment (cwperks) [reply config/opensearch_dashboards.yml:290]: Added possible values as a comment in this file.

File: config/opensearch_dashboards.yml
Comment (peternied) [inline config/opensearch_dashboards.yml:286]: @BionIT @ZilongX This configuration introduces a potential security risk by implying a global setting for certificate authorities across all data sources. This approach might not meet sysadmins' expectations, who typically require the ability to specify distinct certificate authorities for each data source for enhanced security. Could we consider revising this design to allow each data source to be securely and separately configured, ensuring attributes such as the certificate authority are isolated?

Comment (cwperks) [reply config/opensearch_dashboards.yml:286]: @peternied I have called that out on the corresponding issue here: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4524#issuecomment-1989323830

These settings would explicitly be for a common config if TLS is common between the attached data sources. TLS per data source would require UX when configuring a data source.

File: src/plugins/data_source/server/client/client_config.ts
Comment (xinruiba) [inline src/plugins/data_source/server/client/client_config.ts:69]: Looks like we only need `rawConfig.ssl.certificateAuthorities` here.
Maybe we can only pass that value as the parameter of this function instead of pass the whole config with an any type. What do you think?

File: src/plugins/data_source/server/client/client_config.ts
Comment (xinruiba) [inline src/plugins/data_source/server/client/client_config.ts:72]: Nit:
For me, it's more readable to not using abbreviation.
addCAs ----> addCertificateAuthorities
Any concerns?

Comment (cwperks) [reply src/plugins/data_source/server/client/client_config.ts:72]: I chose the abbreviation since its consistent with how the opensearch client can be configured [here](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/opensearch/opensearch_config.ts#L364-L425), but I can update this to not use an abbreviation. CA is the standard abbr for certificate authority

File: src/plugins/data_source/server/client/client_config.ts
Comment (cwperks) [reply src/plugins/data_source/server/client/client_config.ts:69]: In the future, this method could be updated to add more TLS configuration like is done with the client for [default opensearch](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/opensearch/opensearch_config.ts#L364-L425). Since only certificateAuthorities is supported for the datasource opensearch client, I will update this to only consume certificateAuthorities as a parameter.

File: src/plugins/data_source/server/client/client_config.ts
Comment (cwperks) [reply src/plugins/data_source/server/client/client_config.ts]: Removed undefined from types this method consumes since it will never consume undefined.

File: src/plugins/data_source/server/client/client_config.ts
Comment (cwperks) [reply src/plugins/data_source/server/client/client_config.ts]: You're right, it will always be a list and its not necessary to check if its empty or not since the forEach will not execute if its an empty list. Removed this check.

File: src/plugins/data_source/server/client/client_config.ts
Comment (cwperks) [reply src/plugins/data_source/server/client/client_config.ts]: Changed this to consume only the certificateAuthorities param.

File: src/plugins/data_source/server/client/client_config.ts
Comment (cwperks) [reply src/plugins/data_source/server/client/client_config.ts]: Changed all abbreviations to the long form - certificateAuthorities.

File: src/plugins/data_source/server/client/client_config.ts
Comment (cwperks) [reply src/plugins/data_source/server/client/client_config.ts:46]: Marking this as resolved.

File: config/opensearch_dashboards.yml
Comment (cwperks) [reply config/opensearch_dashboards.yml:286]: @peternied Do you think that this should block the PR or can we add this shared setting to unblock testing with HTTPS in the security-dashboards-plugin? I can file an issue around configured TLS per data-source to track the work. That would either require UX where TLS is configured when adding a data source or possibly a separate config file from `opensearch_dashboards.yml` where you can config your data sources from yml instead of through OSD pages.

Comment (cwperks) [reply config/opensearch_dashboards.yml:286]: @BionIT @ZilongX Is there a plan to configure multiple data-sources through a yml file?

Comment (cwperks) [reply config/opensearch_dashboards.yml:286]: Filed an issue to track datasource-specific TLS configuration: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6226

File: src/plugins/data_source/server/client/client_config.ts
Comment (BionIT) [reply src/plugins/data_source/server/client/client_config.ts]: we can use the https://github.com/opensearch-project/OpenSearch-Dashboards/tree/main/src/plugins/data_source/server/util to store commonly used utils then add test there.

Comment (cwperks) [reply src/plugins/data_source/server/client/client_config.ts]: Done. I extracted this to a util and added tests specifically for the util to cover 4 scenarios:

1. List of paths
2. Empty list
3. Undefined
4. Single entry

Let me know what you think.

**Comment by cwperks** [comment]
@BionIT @ZilongX I have replied to all PR feedback, can you take another look when you have a chance?

**Review by BionIT — APPROVED** [review]

**Comment by cwperks** [comment]
I will open a manual backport

**Comment by cwperks** [comment]
@ZilongX @BionIT Created a manual backport to 2.x for this here: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6244

**Pull request merged**
